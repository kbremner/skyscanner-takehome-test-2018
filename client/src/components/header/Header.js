import React, { Component } from 'react';
import { connect } from 'react-redux';
import { searchForPlaces, clearPlaces, placeSelected, search } from '../../store/actions';
import { places, origin, destination, numTravellers, cabinClass } from '../../store/selectors';
import BpkLargeLongArrowRight from 'bpk-component-icon/lg/long-arrow-right';
import BpkText from 'bpk-component-text';
import BpkButton from 'bpk-component-button';
import { withDefaultProps } from 'bpk-react-utils';
import PropTypes from '../../types';
import PlaceAutocomplete from './PlaceAutocomplete';
import './Header.scss';

const PlaceText = withDefaultProps(BpkText, { textStyle: 'xxl' });

export class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editable: false
    };
  }

  handleSearch = () => {
    this.setState(() => ({ editable: false }));
    this.props.search();
  }

  renderEditable() {
    const { origin, destination, places, searchForPlaces, clearPlaces, placeSelected } = this.props;
    return (<div className="header__options">
      <PlaceAutocomplete
        name="fromPlace"
        defaultValue={origin}
        places={places}
        onSearchForPlaces={searchForPlaces}
        onClearPlaces={clearPlaces}
        onPlaceSelected={placeId => placeSelected("fromPlace", placeId)}
      />
      <PlaceAutocomplete
        name="toPlace"
        defaultValue={destination}
        places={places}
        onSearchForPlaces={searchForPlaces}
        onClearPlaces={clearPlaces}
        onPlaceSelected={placeId => placeSelected("toPlace", placeId)}
      />
      <BpkButton className="header__submit-btn" onClick={this.handleSearch}>Go!</BpkButton>
      <BpkButton
        secondary
        className="header__close-btn"
        onClick={() => this.setState(() => ({ editable: false }))}
      >
        Close
      </BpkButton>
    </div>);
  }

  renderNonEditable() {
    const { origin, destination, numTravellers, cabinClass } = this.props;
    return (<div
      aria-label="edit search options"
      tabIndex={0}
      role="button"
      onKeyPress={({ key }) => key === 'Enter' ? this.setState(() => ({ editable: true })) : null}
      onClick={() => this.setState(() => ({ editable: true }))}
    >
      <div className="header__route">
        <PlaceText className="header__place">{origin}</PlaceText>
        <BpkLargeLongArrowRight className="header__arrow" />
        <PlaceText className="header__place">{destination}</PlaceText>
      </div>
     <BpkText className="header__details">{numTravellers} travellers, {cabinClass}</BpkText>
    </div>);
  }

  render() {
    const { editable } = this.state;
    return (<div className="header">
      {editable && this.renderEditable()}
      {!editable && this.renderNonEditable()}
    </div>);
  }
}

Header.propTypes = {
  origin: PropTypes.string.isRequired,
  destination: PropTypes.string.isRequired,
  numTravellers: PropTypes.number.isRequired,
  cabinClass: PropTypes.cabinClass.isRequired,
  places: PropTypes.arrayOf(PropTypes.place).isRequired,
  searchForPlaces: PropTypes.func.isRequired,
  clearPlaces: PropTypes.func.isRequired,
  placeSelected: PropTypes.func.isRequired,
  search: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  return {
    places: places(state),
    origin: origin(state),
    destination: destination(state),
    numTravellers: numTravellers(state),
    cabinClass: cabinClass(state)
  }
}

const mapDispatchToProps = dispatch => {
  return {
    searchForPlaces: (query) => dispatch(searchForPlaces(query)),
    clearPlaces: () => dispatch(clearPlaces()),
    placeSelected: (name, value) => dispatch(placeSelected(name, value)),
    search: () => dispatch(search())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Header);
