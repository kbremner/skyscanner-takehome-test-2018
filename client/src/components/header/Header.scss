@import "~bpk-mixins";

.header {
  padding: $bpk-spacing-sm $bpk-spacing-sm;
  background-color: $bpk-color-blue-500;
}

.header__route {
  display: flex;
  align-items: center;
}

.header__place, .header__details {
  color: $bpk-color-white;
}

.header__arrow {
  fill: $bpk-color-white;
  margin: $bpk-spacing-xs $bpk-spacing-xs;
}

.header__options .bpk-autosuggest__container, .header__submit-btn {
  margin: $bpk-spacing-sm $bpk-spacing-sm;
}