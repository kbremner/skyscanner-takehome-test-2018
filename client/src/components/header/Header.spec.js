import React from 'react';
import { shallow } from 'enzyme';
import { Header } from './Header';

const defaultProps = () => ({
  origin: 'test-origin',
  destination: 'test-destination',
  numTravellers: 4,
  cabinClass: 'first',
  places: [],
  searchForPlaces: jest.fn(),
  clearPlaces: jest.fn(),
  placeSelected: jest.fn(),
  search: jest.fn()
});

const setup = (overrideProps = {}) => {
  const props = {
    ...defaultProps(),
    ...overrideProps
  };
  const wrapper = shallow(<Header {...props} />);
  return { wrapper, props };
};

describe('Header', () => {
  it('renders non-editable view', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders editable view when non-editable view is clicked', () => {
    const { wrapper } = setup();
    wrapper.find({ role: 'button' }).simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('renders editable view when press enter on non-editable view', () => {
    const { wrapper } = setup();
    wrapper.find({ role: 'button' }).simulate('keyPress', { key: 'Enter' });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders non-editable view when close button is clicked', () => {
    const { wrapper } = setup();
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find('.header__close-btn').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('renders non-editable view when go button is clicked', () => {
    const { wrapper } = setup();
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find('.header__submit-btn').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('triggers search when go button is clicked', () => {
    const { wrapper, props } = setup();
    const { search } = props;
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find('.header__submit-btn').simulate('click');
    expect(search).toHaveBeenCalledTimes(1);
  });

  it('updates fromPlace when from place is selected', () => {
    const placeId = Symbol('test-place-id');
    const { wrapper, props } = setup();
    const { placeSelected } = props;
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find({ name: 'fromPlace' }).simulate('placeSelected', placeId);
    expect(placeSelected).toHaveBeenCalledTimes(1);
    expect(placeSelected).toHaveBeenCalledWith('fromPlace', placeId);
  });

  it('clears places when triggered from from place autocomplete', () => {
    const { wrapper, props } = setup();
    const { clearPlaces } = props;
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find({ name: 'fromPlace' }).simulate('clearPlaces');
    expect(clearPlaces).toHaveBeenCalledTimes(1);
  });

  it('calls places search when triggered from from place autocomplete', () => {
    const value = Symbol('test-value');
    const { wrapper, props } = setup();
    const { searchForPlaces } = props;
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find({ name: 'fromPlace' }).simulate('searchForPlaces', value);
    expect(searchForPlaces).toHaveBeenCalledTimes(1);
    expect(searchForPlaces).toHaveBeenCalledWith(value);
  });

  it('updates toPlace when from place is selected', () => {
    const placeId = Symbol('test-place-id');
    const { wrapper, props } = setup();
    const { placeSelected } = props;
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find({ name: 'toPlace' }).simulate('placeSelected', placeId);
    expect(placeSelected).toHaveBeenCalledTimes(1);
    expect(placeSelected).toHaveBeenCalledWith('toPlace', placeId);
  });

  it('clears places when triggered from to place autocomplete', () => {
    const { wrapper, props } = setup();
    const { clearPlaces } = props;
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find({ name: 'toPlace' }).simulate('clearPlaces');
    expect(clearPlaces).toHaveBeenCalledTimes(1);
  });

  it('calls places search when triggered from to place autocomplete', () => {
    const value = Symbol('test-value');
    const { wrapper, props } = setup();
    const { searchForPlaces } = props;
    wrapper.find({ role: 'button' }).simulate('click');
    wrapper.find({ name: 'toPlace' }).simulate('searchForPlaces', value);
    expect(searchForPlaces).toHaveBeenCalledTimes(1);
    expect(searchForPlaces).toHaveBeenCalledWith(value);
  });
});
