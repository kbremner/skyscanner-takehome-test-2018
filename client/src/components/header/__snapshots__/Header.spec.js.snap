// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders editable view when non-editable view is clicked 1`] = `
<div
  className="header"
>
  <div
    className="header__options"
  >
    <PlaceAutocomplete
      defaultValue="test-origin"
      name="fromPlace"
      onClearPlaces={[MockFunction]}
      onPlaceSelected={[Function]}
      onSearchForPlaces={[MockFunction]}
      places={Array []}
    />
    <PlaceAutocomplete
      defaultValue="test-destination"
      name="toPlace"
      onClearPlaces={[MockFunction]}
      onPlaceSelected={[Function]}
      onSearchForPlaces={[MockFunction]}
      places={Array []}
    />
    <BpkButton
      className="header__submit-btn"
      destructive={false}
      disabled={false}
      featured={false}
      href={null}
      iconOnly={false}
      large={false}
      link={false}
      onClick={[Function]}
      secondary={false}
      submit={false}
    >
      Go!
    </BpkButton>
    <BpkButton
      className="header__close-btn"
      destructive={false}
      disabled={false}
      featured={false}
      href={null}
      iconOnly={false}
      large={false}
      link={false}
      onClick={[Function]}
      secondary={true}
      submit={false}
    >
      Close
    </BpkButton>
  </div>
</div>
`;

exports[`Header renders editable view when press enter on non-editable view 1`] = `
<div
  className="header"
>
  <div
    className="header__options"
  >
    <PlaceAutocomplete
      defaultValue="test-origin"
      name="fromPlace"
      onClearPlaces={[MockFunction]}
      onPlaceSelected={[Function]}
      onSearchForPlaces={[MockFunction]}
      places={Array []}
    />
    <PlaceAutocomplete
      defaultValue="test-destination"
      name="toPlace"
      onClearPlaces={[MockFunction]}
      onPlaceSelected={[Function]}
      onSearchForPlaces={[MockFunction]}
      places={Array []}
    />
    <BpkButton
      className="header__submit-btn"
      destructive={false}
      disabled={false}
      featured={false}
      href={null}
      iconOnly={false}
      large={false}
      link={false}
      onClick={[Function]}
      secondary={false}
      submit={false}
    >
      Go!
    </BpkButton>
    <BpkButton
      className="header__close-btn"
      destructive={false}
      disabled={false}
      featured={false}
      href={null}
      iconOnly={false}
      large={false}
      link={false}
      onClick={[Function]}
      secondary={true}
      submit={false}
    >
      Close
    </BpkButton>
  </div>
</div>
`;

exports[`Header renders non-editable view 1`] = `
<div
  className="header"
>
  <div
    aria-label="edit search options"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="header__route"
    >
      <withDefaultProps(BpkText)
        className="header__place"
      >
        test-origin
      </withDefaultProps(BpkText)>
      <Component
        className="header__arrow"
      />
      <withDefaultProps(BpkText)
        className="header__place"
      >
        test-destination
      </withDefaultProps(BpkText)>
    </div>
    <BpkText
      bold={false}
      className="header__details"
      tagName="span"
      textStyle="base"
    >
      4
       travellers, 
      first
    </BpkText>
  </div>
</div>
`;

exports[`Header renders non-editable view when close button is clicked 1`] = `
<div
  className="header"
>
  <div
    aria-label="edit search options"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="header__route"
    >
      <withDefaultProps(BpkText)
        className="header__place"
      >
        test-origin
      </withDefaultProps(BpkText)>
      <Component
        className="header__arrow"
      />
      <withDefaultProps(BpkText)
        className="header__place"
      >
        test-destination
      </withDefaultProps(BpkText)>
    </div>
    <BpkText
      bold={false}
      className="header__details"
      tagName="span"
      textStyle="base"
    >
      4
       travellers, 
      first
    </BpkText>
  </div>
</div>
`;

exports[`Header renders non-editable view when go button is clicked 1`] = `
<div
  className="header"
>
  <div
    aria-label="edit search options"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="header__route"
    >
      <withDefaultProps(BpkText)
        className="header__place"
      >
        test-origin
      </withDefaultProps(BpkText)>
      <Component
        className="header__arrow"
      />
      <withDefaultProps(BpkText)
        className="header__place"
      >
        test-destination
      </withDefaultProps(BpkText)>
    </div>
    <BpkText
      bold={false}
      className="header__details"
      tagName="span"
      textStyle="base"
    >
      4
       travellers, 
      first
    </BpkText>
  </div>
</div>
`;
